Sprint two team two development

DavidOstlund opened the pull request:
We have added the following to the DateTime class:
comparator
classmethod: fromisoformat
instance method: isoformat

And tests for the DateTime class.

Coverage has not yet been looked at.

R055A reviewed (changes requested):
To start with, it does not build in either IntelliJ nor with the ant java command. 

Error in IntelliJ: separator cannot be resolved to a variable (line 179 in DateTime.java). Same error with ant java and build.sh.

My advice is to declare String seperator just before the switch statement, of which has too little cases, and to remove the additions made to build.xml and build-java.xml and maybe the .sh files as the ant java-test command returns the error: Could not find or load main class org.junit.platform.console.ConsoleLauncher.

The .gitignore additions can also be simplified to apache-ant-1.10.7

Until the build fail is fixed cannot continue the review

nguyenluc99 commented on python/common/org/python/stdlib/datetime/DateTime.java:
```diff
@@ -43,186 +48,304 @@
+        switch(isoString.typeName()) {
+            case "str":
+                org.python.types.Str t = (org.python.types.Str) isoString;
+                separator = (String) t.toJava();
```
Define seperator as a String here

R055A reviewed (commented):
Thanks for fixing the build fail bug.

# Junit testing with code coverage

* All testing with code coverage is conducted in IntelliJ. 
* The test file is moved to an appropriate package for this.
* Junit4 imports are changed to junit5. 
* Java version 1.8 is used. 

## All tests

When running all tests for DateTime.java together once, all test cases pass except for test_weekday().

When repeating this testing 100 times in a total time of 633 ms, the results are 591 passed out of 700. Many of these fails are caused by a slight indifference between milliseconds in test and tested code. 

The code coverage results are 18 of the 18 (100%) methods, 157 of the 175 (89%) lines and 12 of the 50 (24%) branches tested. This can be observed in the attached image with green denoting full coverage, yellow denoting partial coverage and red denoting zero coverage. 

## Individual tests

#### test_creation()

When running this test individually it passes all 100 times. The code coverage results are 12 of the 18 (66%) methods, 82 of the 175 (46%) lines and 8 of the 32 (25%) branches.

This is very good. There is also the info argument to consider in testing. 

Also, the (org.python.types.Str) cast is redundant in each of the test and the method being tested. 

#### test_comparator()

When running this test individually it passes all 100 times. The code coverage results are 12 of the 18 (66%) methods, 64 of the 175 (36%) lines and 1 of the 20 (5%) branches. The HashMap explicit type argument is redundant and can be replaced with <>. 

I'm entirely sure what this method is as it is not listed in the Python datetime.datetime library. I assume that it is the extended comparison operation implementation. I also assume it should be returning a boolean value and not one of the DateTime() instances being compared. I also do not think that this was what was intended to be implemented, however. What is more likely intended is for implementation of relational Python operations, such as datetime1 < datetime2 is True or False, datetime1 == datetime2 is True or False, etc. 

#### test_today()

When running this test individually it passes 99 of the 100 times. The code coverage results are 13 of the 18 (72%) methods, 84 of the 175 (48%) lines and 0 of the 31 (0%) branches. 

The single fail of the 100 tests is because of a difference in 14 milliseconds, which just means there is a slight difference in the execution of the test code and the code being tested. 

A potential solution could be to disregard the microseconds in the testing as the test is for the current day, or to accept a difference in microseconds, or even seconds within a reasonable limit, such as +/- 1.

Furthermore, Python print(datetime.datetime.today()) returns a string in the format: yyyy-MM-dd HH:mm:ss.SSSSSS if it includes microseconds, whereas the string format you are returning has only two milliseconds. The test argument should also be second in the assertEquals() method.

Each argument given to the Object[] array args at line 167 could be in a separate line for readability. The code also extends beyond the line limit (provided in IntelliJ) so the arguments should be on separate lines.

#### test_date()

When running this test individually it passes all 100 times. The code coverage results are 13 of the 18 (72%) methods, 67 of the 175 (38%) lines and 0 of the 18 (0%) branches. 

Each argument given to the Object[] array args at line 158 could be in a separate line for readability. The code also extends beyond the line limit so the arguments should be on separate lines.

#### test_weekday()

When running this test individually it passes 0 of the 100 times. The code coverage results are 13 of the 18 (72%) methods, 79 of the 175 (45%) lines and 0 of the 18 (0%) branches. 

The test result never changes from day to day. This is because in the method being tested, at lines 272, 276 and 280, this.year attribute is being set to each of year, month and day arguments being entered to the GregorianCalendar() method. 

The test argument should also be second in the assertEquals() method as it can lead to confusion when reading the results of the failed test.

#### test_isoformat()

When running this test individually it passes 89 of the 100 tests. The code coverage results are 13 of the 18 (72%) methods, 79 of the 175 (45%) lines and 0 of the 18 (0%) branches. 

The 11 fails of the 100 tests is because of a slight indifference between microseconds, which just means there is a slight difference in the execution of the test code and the code being tested.

A potential solution could be to disregard the microseconds in the testing as the test is for the current day, or to accept a difference in microseconds, or even seconds within a reasonable limit such as +/- 1. However, testing the current time is not necessary for datetime().isoformat(); the test can be hard coded. 

Datetime().isoformat() also accepts the format without microseconds in Python, but the correct formatting of microseconds should also be included in any test, which is 6 digits if they are included, no more no less, otherwise 0 if not included, whereas there are only 2 microseconds digits in the test format. There is the option of just having milliseconds, but then the number of digits must be 3. The testing should also include cases where the utcoffset() does not return None.

Thank you for referencing me in the isoformat() method. However, the switch statement is unnecessary with such few cases.

#### test_fromisoformat()

When running this test individually it passes all 100 times. The code coverage results are 13 of the 18 (72%) methods, 93 of the 175 (53%) lines and 3 of the 39 (7%) branches.

Python print(datetime.datetime.fromisoformat()) returns strings of multiple different formats, whereas only one format is being tested. The format being tested will also return a ValueError with the message: Invalid isoformat string: '<String>'. Refer to the datetime.datetime library: [https://docs.python.org/3/library/datetime.html](https://docs.python.org/3/library/datetime.html). 

## Missing testing

There are multiple methods and other functionality not directly tested, such as min(), max(), and other attributes, for example. \__str\__() is involved in many tests, of which shows that the attributes are correct when printed as a string, but neither /__str/__() or any attribute being printed by this method are tested specifically in separate test cases. If there is ever a bug in the /__str/__() method then it will fail all existing test cases aside for test_comparator() and test_weekday() making it difficult to determine where the bug causing these tests to fail is. 

There is little to no consideration for testing different input types and different input value ranges for each tested method.

# Code review

* There is no module or method javadoc anywhere in either DateTime.java or DateTimeTest.java. 
* There are unused files from Team One's Sprint Two development included in the pull request.
    * Strings for error messages are being used where Team One's development includes exisitng enums for these.
    * The files from Team One's development are not updated with recent edits and could potentially cause a merge conflict.
        * It is possible to git add and git commit unstaged files individually by specifying the files after each git command.
* @expectedFailure should not be annotated to any of the DateTime tests in tests.stdlib.test_datetime
    * if the tests are failing then it is because they are not returning the correct error message.
* Flake8 && ant checkstyle returned multiple errors for test_datetime.py
    * although these errors don't directly relate to your edits, the file is included in this pull request

Congratulations on the good work done!

![Screenshot_2020-09-29 Coverage Report DateTime(1)](https://user-images.githubusercontent.com/14985050/94440142-c9449200-01fd-11eb-9570-6cb995ed4737.png)

ponterik commented:
I pushed changes for the methods I´ve been working on to this pr, see below with some added comments
There will be one more push with more changes from David soon.

**today()**
Regarding datetime.today() ours returns yyyy-MM-dd HH:mm:ss.SSSSSS but we chose to test on the format yyyy-MM-dd HH:mm:ss.SS since the last four milliseconds always made us fail the test.
Moved the requested code to separate rows.
**date()**

Moved the requested code to separate lines.

**isoformat()**
We do not have to consider the case where utcoffset() does not return none since tzinfo can be assumed to be none (mentioned in the spec for this sprint). Fixed some bugs with the milliseconds.

In general
Expanded the tests to test more scenarios.
Switched place of the expected arguements in many places to the correct position

DavidOstlund commented:
Here is a comment about my changes.
For all functions I ahve assumed that tzinfo is always None, as has been approved in the instructions.
Most functions also don't catch wrong inputs purely because of time restraints.

#### test_datetime.py
Fixed flake8 and removed ExpectedError

#### DateTime class
Fixed some error messages
I've also added tests for creation of the DateTime class with wrong sizes on the inputs.

#### fromisoformat()
Added support for the other types of input formats.
(Had accidentally looked fromisoformat function for Date at first)

#### comparison tunctions __eq__(), __gt__(), __ge__(), __lt__(), __le__()
I have now implemented the correct kinds of comparison functions.
I am well aware that the functions are very alike, and with more time it would probably be possible to factor something out.
I also know that it is not the most beautiful solution, but I've found working with the Python objects to be a bit clunky and this was the best solution I was able find with the time available.
The __lt__() and __le__() functions are using __gt__() and __ge__() which is why there are so few tests for them.

nguyenluc99 commented:
When I run the test, it say: java: package org.junit does not exist
Does anyone know how to fix this?

R055A commented:
> 
> 
> When I run the test, it say: java: package org.junit does not exist
> Does anyone know how to fix this?

Move the DateTimeTest.java file to the same package the DateTest.java file is in, package: org.test.python.stdlib.datetime. Replace the junit4 imports in DateTimeTest.java with junit5 imports, such as the junit5 imports used in DateTest.java

merryaiiit reviewed (approved):
I have now gone through the previous commented report and how now confirmed that the tests are passing and that compatibility with IntelliJ and junit5 is working.

merryaiiit commented:
Running the code with code coverage us:
100% Class coverage
100% Method coverage
93% Line coverage

This means that it passes our review rules (minimum of 80%)

Well done Team Two!

R055A reviewed (commented):
I have run all of the tests together and individually and can confirm that they all pass with improved code coverage. The line and branch coverage when running all tests is now 93% and 53%, respectively, which is over twice the branch coverage as before. 

It is a great improvement from the last review with many significant fixes and many more test cases.

As for individual test cases, I did not have enough to review all of them in too much detail. I did find that the weekday method is still not passing although the tests you created are returning the correct results now for what they do test. When testing a date such as 0001-01-01 Python will return 0 for the weekday, whereas the DateTime.weekday() method implementation will return 5. A solution to this can be found in the Date.weekday() implementation.

In general, the code is quite untidy and the convention looks inconsistent in DateTime.java and DateTimeTest.java, such as line spacing, naming, javadoc, lines too long, too many indentations in some methods, etc. 

There are a lot of strings, many of these being repeated. Enums can be used in place of these, which would improve maintainability and readability with all the strings in one place and only one of each, and have the code looking cleaner and with shorter lines in many places.

So, if you can fix the weekday() method then that would be good. Otherwise, it is passing for what I can see
